### PR TITLE
make Http2SolrClient max outstanding requests configurable via system property

### PR DESCRIPTION
**Pull request by magibney**
See https://issues.apache.org/jira/browse/SOLR-16932

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 1]
```diff
@@ -1016,7 +1017,8 @@ private static class AsyncTracker {
 
     int getMaxRequestsQueuedPerDestination() {
       // comfortably above max outstanding requests
-      return MAX_OUTSTANDING_REQUESTS * 3;
+      // constant above MAX_OUTSTANDING_REQUESTS, not linear function
```
Comment (hiteshk25) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:1020]: what is the motivation to change this?

**Review by hiteshk25 — APPROVED** [review]
LGTM

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 1 above)
Comment (magibney) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:1020]: This is the [relevant section of the jetty code](https://github.com/eclipse/jetty.project/blob/jetty-10.0.15/jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java#L312-L333). The queue whose size is defined by this `maxRequestsQueuedPerDestination()` is the `exchanges` Queue, added to on the first linked line (`enqueue(exchanges, exchange)`). This initial queue is basically a buffer -- all requests hit this queue _before_ acquiring the "send" Semaphore from AsyncTracker.

It is then the `requestNotifier.notifyQueued(request)` line that acquires (blocking) a "send" permit from AsyncTracker, before proceeding with the actual "send" logic immediately afterward (`send()`, which drains the destination queue).

As I look at this, the way it works seems a bit arbitrary actually, and doesn't exactly work the way one might think. But in any event, I'm fairly certain that it's misleading for this queue to be related to `maxOutstandingRequests` at all. In fact, to the extent that they are related, it's that the queue must be sufficiently large to buffer requests until they can be sent. Increasing `maxOutstandingRequests` should thus increase throughput to the extent that a very small `maxRequestsQueuedPerDestination` should suffice.

I think the right way to think about `maxRequestsQueuedPerDestination` is: How many requests (per destination) is it reasonable to buffer, figuring "I'll send this as soon as I can" vs failing fast and sending an error code to the client indicating that we're too backed up right now. I could even see just making this a static value (e.g., the 3k that it has effectively been up to this point). I think I was thrown by the "comfortably above max outstanding requests" comment though, implying some relationship there. Tbh I can't think of any reason why this should even necessarily be "above" max outstanding requests, let alone "comfortably above" ...

Comment (magibney) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:1020]: In any event, leaving the per-destination queue size determined as 3x linear function of max outstanding requests really doesn't make sense, and I think that would make the queue absurdly long (esp. considering that, in practice, and even with the artificial bandwidth limitation from `MAX_OUTSTANDING_REQUESTS`, we evidently haven't been overflowing these per-destination queues at their current 3k size.

Comment (magibney) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:1020]: Ok, I'm glad I dug into this to understand it better -- will stand us in good stead if we have any further debugging we need to do around this.

But to be clear, I now think we should just make `maxRequestsQueuedPerDestination ` a static value, 3k. It's wrt per-destination, so shouldn't have the same scaling concerns as `MAX_OUTSTANDING_REQUESTS`, which is per-Http2SolrClient, across all destinations.

**Pull request merged**